core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    protected int getSessionsCount(List<Map<String, Object>> nodes) {
+        int availableSessions = 0;
+        for (Map<String, Object> node : nodes) {
+            List<Map<String, Object>> slots = (List<Map<String, Object>>) node.get("slots");
+            for (Map<String, Object> slot : slots) {
+                if (slot.get("session") == null) {
+                    availableSessions++;
+                }
+            }
+        }
+        return availableSessions;
+    }
+
+    public class Holder {
+        public RemoteWebDriver driver;
+        public Long time;
+
+        public void setDriver(RemoteWebDriver driver) {
+            this.driver = driver;
+        }
+
+        public void setTime(Long time) {
+            this.time = time;
+        }
+
+        public RemoteWebDriver getDriver() {
+            return this.driver;
+        }
+
+        public Long getTime() {
+            return this.time;
+        }
+
+        public Holder(RemoteWebDriver driver, Long time) {
+            this.driver = driver;
+            this.time = time;
+        }
+    }
+
+    public ProtocolResponse getProtocolOutput(String url, Metadata metadata) throws Exception {
+        RemoteWebDriver driver;
+        while ((driver = getDriver()) == null) {}
+        try {
+            // This will block for the page load and any
+            // associated AJAX requests
+            driver.get(url);
+
+            String u = driver.getCurrentUrl();
+
+            // call the filters
+            ProtocolResponse response = filters.filter(driver, metadata);
+            if (response != null) {
+                return response;
+            }
+
+            // if the URL is different then we must have hit a redirection
+            if (!u.equalsIgnoreCase(url)) {
+                byte[] content = new byte[] {};
+                Metadata m = new Metadata();
+                m.addValue(HttpHeaders.LOCATION, u);
+                return new ProtocolResponse(content, 307, m);
+            }
+
+            // if no filters got triggered
+            byte[] content = driver.getPageSource().getBytes();
+            return new ProtocolResponse(content, 200, new Metadata());
+
+        } catch (Exception e) {
+            if (e.getMessage() != null) {
+                if ((e.getMessage().contains("ERR_NAME_NOT_RESOLVED")
+                        || e.getMessage().contains("ERR_CONNECTION_REFUSED")
+                        || e.getMessage().contains("ERR_CONNECTION_CLOSED")
+                        || e.getMessage().contains("ERR_SSL_PROTOCOL_ERROR")
+                        || e.getMessage().contains("ERR_CONNECTION_RESET")
+                        || e.getMessage().contains("ERR_SSL_VERSION_OR_CIPHER_MISMATCH")
+                        || e.getMessage().contains("ERR_ADDRESS_UNREACHABLE"))) {
+                    LOG.info(
+                            "Exception is of webpage related hence continuing with the driver and adding it back to queue");
```

Comment:
"Encountered a webpage related exception: '{}'. Adding the driver back to the queue.